### PR TITLE
feat: add batching capability for segmenters

### PR DESCRIPTION
**Pull request by JoanFM**

### CODE REVIEW

File: tests/integration/crud/advanced/yaml/custom_executor.py [hunk 1]
```diff
@@ -18,6 +18,7 @@ def __init__(self, *args, **kwargs):
             '\s*([^{0}]+)(?<!\s)[{0}]*'.format(''.join(set(punct_chars)))
         )
 
+    @single
```
Comment (JoanFM) [inline tests/integration/crud/advanced/yaml/custom_executor.py:21]:
```suggestion
    @single(flatten_output=False)
```

File: tests/integration/mime/test_segmenter.py [hunk 2]
```diff
@@ -3,14 +3,16 @@
...
+from jina.executors.decorators import single
 from jina.flow import Flow
 from tests import random_docs, validate_callback
 
 cur_dir = os.path.dirname(os.path.abspath(__file__))
 
 
 class DummySegment(BaseSegmenter):
-    def segment(self):
+    @single
```
Comment (JoanFM) [inline tests/integration/mime/test_segmenter.py:14]:
```suggestion
    @single(flatten_output=False)
```

File: tests/unit/drivers/querylang/test_querylang_drivers.py [hunk 3]
```diff
@@ -45,11 +46,13 @@ def random_docs_with_tags():
 
 
 class DummySegmenter(BaseSegmenter):
+    @single
```
Comment (JoanFM) [inline tests/unit/drivers/querylang/test_querylang_drivers.py:49]:
```suggestion
    @single(flatten_output=False)
```

File: tests/unit/drivers/querylang/test_querylang_drivers.py [hunk 4]
```diff
@@ -45,11 +46,13 @@ def random_docs_with_tags():
...
     def segment(self, text, *args, **kwargs):
         return [{'text': 'adasd' * (j + 1), 'tags': {'id': j}} for j in range(10)]
 
 
 class DummyModeIdSegmenter(BaseSegmenter):
+    @single
```
Comment (JoanFM) [inline tests/unit/drivers/querylang/test_querylang_drivers.py:55]:
```suggestion
    @single(flatten_output=False)
```

File: tests/unit/drivers/test_reduce_all_driver.py [hunk 5]
```diff
@@ -27,6 +28,7 @@ def docs():
 
 
 class MockSegmenterReduce(BaseSegmenter):
+    @single
```
Comment (JoanFM) [inline tests/unit/drivers/test_reduce_all_driver.py:31]:
```suggestion
    @single(flatten_output=False)
```

File: tests/unit/drivers/test_segmenter_driver.py [hunk 6]
```diff
@@ -14,6 +15,7 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.required_keys = {'text'}
 
+    @single
```
Comment (JoanFM) [inline tests/unit/drivers/test_segmenter_driver.py:18]:
```suggestion
    @single(flatten_output=False)
```

File: tests/unit/flow/test_flow_merge.py [hunk 7]
```diff
@@ -12,7 +13,8 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self._label = random.random()
 
-    def segment(self):
+    @single
```
Comment (JoanFM) [inline tests/unit/flow/test_flow_merge.py:16]:
```suggestion
    @single(flatten_output=False)
```

File: tests/unit/flow/test_flow_multimode.py [hunk 8]
```diff
@@ -7,13 +7,15 @@
...
 from jina.executors.indexers.keyvalue import BinaryPbIndexer
+from jina.executors.decorators import single
 from jina.flow import Flow
 from jina.proto import jina_pb2
 
 cur_dir = os.path.dirname(os.path.abspath(__file__))
 
 
 class MockSegmenter(BaseSegmenter):
+    @single
```
Comment (JoanFM) [inline tests/unit/flow/test_flow_multimode.py:18]:
```suggestion
    @single(flatten_output=False)
```

File: jina/drivers/segment.py [hunk 9]
```diff
@@ -19,24 +20,44 @@ def __init__(
...
+        **kwargs,
     ):
         super().__init__(
             executor, method, traversal_paths=traversal_paths, *args, **kwargs
         )
 
     def _apply_all(self, docs: 'DocumentSet', *args, **kwargs):
+        if not docs:
+            self.logger.warning(f'an empty DocumentSet {docs}')
+            return
```
Comment (florian-hoenicke) [inline jina/drivers/segment.py:32]: Interesting, it is saying that some cases are not covered by tests. Is it correct?

Comment (JoanFM) [reply jina/drivers/segment.py:32]: wekk right now this case is also covered in the mixin

**Review by nan-wang — APPROVED** [review]
LGTM👍

**Pull request merged**